### ISSUE TITLE
Provide a Zeek log schema

### ISSUE THREAD

**Issue by ckreibich** [zeek/zeek#1742]
We regularly encounter the need for a given Zeek version/installation to answer the question "What do the logs look like?". Several folks have built their own takes on this in the past, addressing their own requirements, but we have nothing official. It seems that at a minimum...

- It should capture the full set of log types available, the type, name and meaning of each field in that log, as well as their Zeek-level attributes
- It should get reported in a way that's easy to post-process (making JSON a likely contender)
- It needs to be able to capture a specific Zeek installation so that package-specific additions, redefs, etc are accounted for
- It should enable diffing so for example the differences in logs from one version to the next are captured

There's related work:
- Jon's https://github.com/jsiwek/zeek-print-log-info
- [JSON Schema](https://json-schema.org/) provides a way to describe datasets. It won't capture the nuances of Zeek log semantics but offers a starting point we could build on
- Several solutions exist for diffing JSON

This is just a starting point. It'd be great to hear from folks who've wanted this in the past, to understand their requirements.

**Comment by stevesmoot** [comment]
Not immediately addressed by the JSON request - would like to filter and create in markup a la the docs e.g.,
`conn
----
.. list-table::
   :header-rows: 1
   :class: longtable
   :widths: 1 3

   * - Field (Type)
     - Description

   * - ``ts`` (time)
     - This is the time of the first packet.`
(putting markup in markup, may not come through right, but looks ok here in editor).

**Comment by anthonykasza** [comment]
The idea of Zeek generating its own log schemas came from an internal conversation at Corelight.  We thought it would be neat and useful to have Zeek generate its own schema at runtime since log formats and names are applied in a layering fashion and plugins/packages may alter formats.

As all log stream activation happens within `zeek_init` events, one could set a very low priority `zeek_init` event which prints all streams it knows about. Props to @sethhall for the below script.

```
function describe_record(outer_field: string, rf: record_field_table): string_vec
	{
	local lines = string_vec();
	for ( field, meta in rf )
		{
		if ( meta$log )
			{
			if ( /^record / in meta$type_name )
				{
				local record_type = split_string(meta$type_name, / /)[1];
				local new_lines = describe_record(field, record_fields(record_type));
				for ( i in new_lines )
					{
					lines += new_lines[i];
					}
				}
			else
				{
					{
					local field_prefix = outer_field == "" ? "" : cat(outer_field, ".");
					lines += fmt("    \"%s%s\": { \"type\":\"%s\", \"can_be_null\":%s }", field_prefix, field, meta$type_name, meta?$default_val ? "true": "false");
					}
				}
			}
		}
	return lines;
	}

event zeek_init()  &priority=-10000000000
	{
	local log_records: set[string];
	print "{";
	local i = 0;
	for ( id, stream in Log::active_streams )
		{
		++i;
		print fmt("\"%s\" : {", stream$path);
		local rf = record_fields(stream$columns);
		print join_string_vec(describe_record("", rf), ",");
		print fmt("  }%s", i == |Log::active_streams| ? "" : ",");
		}
	print "}";
	}
```

As @stevesmoot raised, sub-records shared across logged records will be need some consideration.

**Comment by ckreibich** [comment]
> As all log stream activation happens within zeek_init events

That's an aspect that gets into thorny terrain quickly ... technically this is a convention, but one that packages thankfully usually follow. But you could have streams created dynamically, and can definitely have ones disabled but present. Depending on your use case you might want to see everything that could possibly be logged given your installation, or only those logs enabled in a given config.

**Comment by chadbrewbaker** [comment]
Looks you can append  "LogAscii::use_json=T"  to external/zeek-testing/tests{ipv6.zeek, m57-long.zeek, m57-short.zeek} to get JSON output. Of course this fails the tests - investigating if the tests do blind string matching or actually parse TSV.

**Comment by ckreibich** [comment]
Two pointers to related technologies, for future reference:
- An [IPFIX schema](https://github.com/DFDLSchemas/IPFIX/blob/master/src/main/resources/org/mitre/ipfix/common.dfdl.xsd) expressed in [DFDL](https://cboblog.typepad.com/cboblog/2008/07/dfdl-data-forma.html)
- The data format templating features in [Netflow v9](https://datatracker.ietf.org/doc/html/rfc3954#page-11)

**Comment by chadbrewbaker** [comment]
I think I am just going to write a direct to Parquet writer. 99% seem to use that for storage anyway. 

I can use the Zed Parquet writer for reference? https://github.com/brimdata/zed/blob/main/cli/zq/command.go

**Comment by chadbrewbaker** [comment]
I actually planned on using the Arrow Parquet writer - can keep it in Arrow too.  Most large processors are using a Parquet backed distributed SQL engine like AWS Athena or DataFusion.

**Comment by ckreibich** [comment]
> If the objective is indeed providing a canonical anchor for how stock Zeek data looks like

Yeah, where "stock" means the log data produced by Zeek in any particular configuration, ideally with clarity on how that configuration resulted. For example, it'd be nice to see that column `community_id` in this particular conn.log over here was provided by the Community ID Zeek package.

The goal is not to produce schema data for any particular ingestion platform or toolchain. That said, if folks have examples of data schema from other domains or how schema data are consumed by popular platforms, feel free to mention them here.

So far it seems that a superset of JSON Schema is a solid option. It's also not particularly hard to envision keeping the rendering modular so that people can provide their own if really needed.

**Comment by ckreibich** [comment]
@chadbrewbaker can you say more re "extra Zeek table of eBPF logs by timestamp"? I'm asking mainly because eBPF is a natural way to expand the cababilities of the [new zeek-agent](https://github.com/zeek/zeek-agent-v2) on Linux.

**Comment by mavam** [comment]
> For example, it'd be nice to see that column `community_id` in this particular conn.log over here was provided by the Community ID Zeek package.

One way to achieve quasi bullet-proof is by hooking into the stream of events via Broker: you always get the schema, followed by the bulk data. If you have a test suite that generates all types of stock data, you have a starting point.

You could go the extra mile and loop over all existing packages in the official package source and run a specific test that spits out a log. This may require a customization point for the package writer, i.e., opt-in, but with the incentive that if they provide such a unit test, their package will be subject to an extension of the "official" schema.

**Comment by ckreibich** [comment]
At long last an update on this one — we now have [this Zeek package](https://github.com/zeek/logschema) to handle schema production, in several formats. There are many additional questions regarding properties that the produced schemas should have (particularly for JSON Schema), and there's for sure no one-size-fits-all for this. So it's best to handle these on a case-by-case basis for the schema format in question, on that package. We'll write up a little more on this in the near future.

I'm going to close this issue at this point since that package provides the basic functionality of reporting what Zeek's logs "look like" via a given filter.
